### ISSUE TITLE
ipv6 connectivity -- managing NDP entries

### ISSUE THREAD

**Issue by nivekuil** [containers/netavark#270]

**Comment by Luap99** [comment]
Moved to netavark since we would need to implement it here.
I assume you want to use this with rootful podman because rootless would be another story.

I think netavark does only ipv6 nat at the moment and I think you want global routeable addresses?
Why would need to manually manage ndp entries? Shouldn't the kernel handle this automatically?

**Comment by mheon** [comment]
I think we do support globally-routable addresses, but I don't know if we documented the "how" of it particularly well, and we definitely aren't testing them in CI on account of not having any to work with.

**Comment by Luap99** [comment]
I do not see any conditionals in the ipv6 iptables code so i think we always NAT them.

**Comment by mheon** [comment]
Well, damn. I know we talked about it, evidently it wasn't implemented.

Should definitely get on this, as it's a regression from CNI's v6 support; @baude PTAL

**Comment by Luap99** [comment]
I don't think cni supports this either

**Comment by mheon** [comment]
We've definitely been instructing people to do this in CNI land for several years. I have to imagine at least some of them were successful?

**Comment by nivekuil** [comment]
> I think netavark does only ipv6 nat at the moment and I think you want global routeable addresses? Why would need to manually manage ndp entries? Shouldn't the kernel handle this automatically?

Vanishingly few hosting providers will give you route the entire ipv6 subnet they assign you. Usually there needs to be
something on your host responding to the upstream NDP solicitations for each pod. Linux can proxy ndp natively but does not automatically know how to respond
to these and needs to be configured manually (see https://github.com/DanielAdolfsson/ndppd/blob/0.2.5/README,
https://yoursunny.com/t/2021/ndpresponder/)

**Comment by nivekuil** [comment]
This can be done with an OCI hook.
```sh
#!/bin/sh
in=$(</dev/stdin)
ip=$(echo $in | jq -r '.annotations | .[\"hatchery.pod.ip\"]')
if=$(/usr/sbin/ip -6 r | awk '/^default/ {printf $5}')
/usr/sbin/ip -6 neigh add proxy $ip dev $if
```

**Comment by nivekuil** [comment]
Yes, rootful. Vanishingly few hosting providers will actually route the entire ipv6 subnet they give you, so usually there needs to be something on your host responding to the upstream NDP solicitations for each pod. Linux does not automatically know how to respond to these and needs to be configured manually (see https://github.com/DanielAdolfsson/ndppd/blob/0.2.5/README, https://yoursunny.com/t/2021/ndpresponder/)

Mar 28, 2022 4:23:32 AM Paul Holzinger ***@***.***>:

**Comment by ruabmbua** [comment]
Is there any update on this? I want to use public routable ipv6 addresses for my containers, but I can not find any documentation on how to do this with podman.

**Comment by Luap99** [comment]
Not NAt-ed global ipv6 address sounds easy enough to implement, contributions welcome.
However my ipv6 knowledge is very limited, if managing NDP entries for this is required than it looks much more complex and I don't think I have time to look into it anytime soon.

**Comment by ruabmbua** [comment]
Got it working by fiddling with network namespaces and interfaces by hand. Should be easy to implement, but e.g. integrating with systemd-networkd for prefix delegation from host interface will be more difficult.

Might start working on a patch for systemd to add an API for container runtimes if it does not exist
